openpgp/keys.go

```diff
@@ -477,16 +498,31 @@ func addSubkey(e *Entity, packets *packet.Reader, pub *packet.PublicKey, priv *p
 			// packets that are in the wrong place (like misplaced 0x13 signatures)
 			// until we get to one that works.  For a test case,
 			// see TestWithBadSubkeySignaturePackets.
+
 			continue
 		}
-		subKey.Sig = sig
-		err = e.PrimaryKey.VerifyKeySignature(subKey.PublicKey, subKey.Sig)
+		err = e.PrimaryKey.VerifyKeySignature(subKey.PublicKey, sig)
 		if err != nil {
-			return errors.StructuralError("subkey signature invalid: " + err.Error())
+			// Non valid signature, so again, no need to abandon all hope, just continue;
+			// make a note of the error we hit.
+			lastErr = errors.StructuralError("subkey signature invalid: " + err.Error())
+			continue
+		}
+		switch sig.SigType {
+		case packet.SigTypeSubkeyBinding:
+			subKey.Sig = sig
+		case packet.SigTypeSubkeyRevocation:
+			subKey.Revocation = sig
 		}
-		break
 	}
-	e.Subkeys = append(e.Subkeys, subKey)
+	if subKey.Sig != nil {
+		e.Subkeys = append(e.Subkeys, subKey)
+	} else {
+		if lastErr == nil {
+			lastErr = errors.StructuralError("Subkey wasn't signed or revoked")
```

Comment:
Error message nit: the key might very well be revoked, even though somehow it wasn't signed.